File: Fortran/Recap/Reduction/scan.f90
!Program to use MPI_Scan to calculate the scan (partial reduction)
!This calculates the sum of all ranks <= the current rank
!So the results should be
!0 => 0
!1 => 0 + 1 = 1
!2 => 0 + 1 + 2 = 3
!3 => 0 + 1 + 2 + 3 = 6
!etc.
PROGRAM scan

  USE mpi
  IMPLICIT NONE

  INTEGER, PARAMETER :: tag = 100

  INTEGER :: rank, recv
  INTEGER :: nproc, ierr

  CALL MPI_Init(ierr)

  CALL MPI_Comm_size(MPI_COMM_WORLD, nproc, ierr)
  CALL MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)

  CALL MPI_Scan(rank, recv, 1, MPI_INTEGER, MPI_SUM, MPI_COMM_WORLD, ierr)
  PRINT "(A,I3,A,I3)","The sum of all ranks up to current on rank ", &
      rank, " is ", recv

  CALL MPI_Finalize(ierr)

END PROGRAM scan

File: C/Recap/Reduction/scan.c
/*Program to use MPI_Scan to calculate the scan (partial reduction)
This calculates the sum of all ranks <= the current rank
So the results should be
0 => 0
1 => 0 + 1 = 1
2 => 0 + 1 + 2 = 3
3 => 0 + 1 + 2 + 3 = 6
etc.*/

#include <stdio.h>
#include <mpi.h>

#define TAG 100

int main(int argc, char ** argv)
{

  int rank, nproc, recv;

  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD, &nproc);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  MPI_Scan(&rank, &recv, 1, MPI_INT, MPI_SUM, MPI_COMM_WORLD);
  printf("The sum of all ranks up to current on rank %3d is %3d\n", rank, recv);

  MPI_Finalize();

}
